File: backend/git_integration.py
import logging
import subprocess
from dataclasses import dataclass
from typing import Optional

import requests

logger = logging.getLogger(__name__)


@dataclass
class GitResult:
    success: bool
    message: str = ""
    stdout: str = ""
    stderr: str = ""


@dataclass
class PRInfo:
    pr_url: str
    pr_number: int
    title: str = ""


class GitIntegration:
    def __init__(self, repo_path: str):
        self.repo_path = repo_path

    def _run_git(self, *args: str, check: bool = False) -> subprocess.CompletedProcess:
        return subprocess.run(
            ["git", *args],
            cwd=self.repo_path,
            capture_output=True,
            text=True,
            check=check,
        )

    def create_branch(self, branch_name: str) -> GitResult:
        result = self._run_git("checkout", "-b", branch_name)
        if result.returncode == 0:
            return GitResult(success=True, message=f"Created branch {branch_name}", stdout=result.stdout)
        return GitResult(success=False, message=result.stderr, stderr=result.stderr)

    def checkout_branch(self, branch_name: str) -> GitResult:
        result = self._run_git("checkout", branch_name)
        if result.returncode == 0:
            return GitResult(success=True, message=f"Switched to {branch_name}", stdout=result.stdout)
        return GitResult(success=False, message=result.stderr, stderr=result.stderr)

    def get_current_branch(self) -> str:
        result = self._run_git("branch", "--show-current")
        return result.stdout.strip()

    def commit_changes(self, message: str) -> GitResult:
        self._run_git("add", "-A")
        result = self._run_git("commit", "-m", message)
        if result.returncode == 0:
            return GitResult(success=True, message="Committed successfully", stdout=result.stdout)
        return GitResult(success=False, message=result.stderr or result.stdout, stderr=result.stderr)

    def push_to_remote(self, remote: str, branch: str) -> GitResult:
        result = self._run_git("push", remote, branch)
        if result.returncode == 0:
            return GitResult(success=True, message="Pushed successfully", stdout=result.stdout)
        return GitResult(success=False, message=result.stderr, stderr=result.stderr)

    def create_pull_request(
        self,
        title: str,
        body: str,
        head_branch: str,
        base_branch: str,
        github_token: str,
        repo_owner: str,
        repo_name: str,
    ) -> PRInfo:
        url = f"https://api.github.com/repos/{repo_owner}/{repo_name}/pulls"
        headers = {
            "Authorization": f"token {github_token}",
            "Accept": "application/vnd.github.v3+json",
        }
        payload = {
            "title": title,
            "body": body,
            "head": head_branch,
            "base": base_branch,
        }
        resp = requests.post(url, json=payload, headers=headers)
        resp.raise_for_status()
        data = resp.json()
        return PRInfo(
            pr_url=data["html_url"],
            pr_number=data["number"],
            title=data.get("title", ""),
        )

    def get_status(self) -> str:
        result = self._run_git("status", "--short")
        return result.stdout.strip()

    def get_diff(self, staged: bool = False) -> str:
        args = ["diff", "--cached"] if staged else ["diff"]
        result = self._run_git(*args)
        return result.stdout

    def get_log(self, count: int = 10) -> str:
        result = self._run_git("log", f"--oneline", f"-{count}")
        return result.stdout.strip()

File: src/modules/gitIntegration.ts
/**
 * Git 集成模块
 *
 * 提供 Git 仓库操作和 GitHub PR 创建功能。
 * 用于 Electron/Node.js 环境，通过 child_process.execSync 执行 git 命令。
 */

import { execSync } from 'child_process'

export interface GitResult {
  success: boolean
  message: string
  stdout: string
  stderr: string
}

export interface PRInfo {
  prUrl: string
  prNumber: number
  title: string
}

export class GitIntegration {
  private readonly repoPath: string

  constructor(repoPath: string) {
    this.repoPath = repoPath
  }

  /**
   * 执行 git 命令并返回结构化结果
   */
  runGit(...args: string[]): GitResult {
    try {
      const stdout = execSync(`git ${args.join(' ')}`, {
        cwd: this.repoPath,
        encoding: 'utf-8',
        timeout: 30000,
      })
      return { success: true, message: '', stdout, stderr: '' }
    } catch (e: any) {
      return {
        success: false,
        message: e.stderr || e.message,
        stdout: e.stdout || '',
        stderr: e.stderr || '',
      }
    }
  }

  /**
   * 创建新分支
   */
  createBranch(name: string): GitResult {
    return this.runGit('checkout', '-b', name)
  }

  /**
   * 切换到指定分支
   */
  checkoutBranch(name: string): GitResult {
    return this.runGit('checkout', name)
  }

  /**
   * 获取当前分支名称
   */
  getCurrentBranch(): GitResult {
    return this.runGit('rev-parse', '--abbrev-ref', 'HEAD')
  }

  /**
   * 暂存所有更改并提交
   */
  commitChanges(message: string): GitResult {
    const addResult = this.runGit('add', '-A')
    if (!addResult.success) return addResult
    return this.runGit('commit', '-m', JSON.stringify(message))
  }

  /**
   * 推送到远程仓库
   */
  pushToRemote(remote: string, branch: string): GitResult {
    return this.runGit('push', remote, branch)
  }

  /**
   * 通过 GitHub API 创建 Pull Request
   */
  async createPullRequest(
    title: string,
    body: string,
    headBranch: string,
    baseBranch: string,
    githubToken: string,
    repoOwner: string,
    repoName: string
  ): Promise<PRInfo> {
    const url = `https://api.github.com/repos/${repoOwner}/${repoName}/pulls`
    const response = await fetch(url, {
      method: 'POST',
      headers: {
        Authorization: `token ${githubToken}`,
        'Content-Type': 'application/json',
        Accept: 'application/vnd.github.v3+json',
      },
      body: JSON.stringify({ title, body, head: headBranch, base: baseBranch }),
    })

    if (!response.ok) {
      const errorText = await response.text()
      throw new Error(`GitHub API error ${response.status}: ${errorText}`)
    }

    const data = await response.json()
    return {
      prUrl: data.html_url,
      prNumber: data.number,
      title: data.title,
    }
  }

  /**
   * 获取工作区状态
   */
  getStatus(): GitResult {
    return this.runGit('status', '--porcelain')
  }

  /**
   * 获取差异内容
   */
  getDiff(staged = false): GitResult {
    if (staged) {
      return this.runGit('diff', '--cached')
    }
    return this.runGit('diff')
  }

  /**
   * 获取提交日志
   */
  getLog(count = 10): GitResult {
    return this.runGit('log', `--oneline`, `-n`, String(count))
  }
}
